### ISSUE TITLE
[BUG] Keydb multi-master replication issues

### ISSUE THREAD

**Issue by iacovosb** [Snapchat/KeyDB#440]
**Describe the bug**

Hello, and thank you for accepting my inquiry.
I have set up four keydb nodes, in a multimaster active-replica configuration.
I am using a Load Balancer in front of the keydb nodes, and I am directing write operations to keydb-1, read operations to keydb-2, and keydb-3 keydb-4 are standing by in case of a failure. 
Replication is set up in a mesh topology, where each of the nodes is a replicaof all the other nodes.
Problem is that I have a different count of keys an the keydb-1 (writer node), than in the keydb-2 (reader node)

**Environment Information**
***Host Information***
```yaml
Host OS: 4 Debian 10 hosts
CPU: 52 cores per host
Memory: 512Gb per host
```

***Keydb Information***
```yaml
Containerized environment: Docker
docker_networking_mode:  host
Image: eqalpha/keydb:x86_64_v6.3.0
```

***Configuration Information***
Configuration file on all hosts:
```yaml
# General
maxclients 50000

# debug (a lot of information, useful for development/testing)
# verbose (many rarely useful info, but not a mess like the debug level)
# notice (moderately verbose, what you want in production probably)
# warning (only very important / critical messages are logged)
loglevel notice
logfile ""

# Persistency
save ""

# Security
protected-mode yes
requirepass {{ mypass }}
masterauth {{ mypass }}

# Network
port 6379
timeout 3600

# Slow Log
slowlog-log-slower-than 1000000
slowlog-max-len 128

# client buffer limits 
client-output-buffer-limit pubsub 0 0 0

# The bigger the replication backlog, the longer the time the replica can be
# disconnected and later be able to perform a partial resynchronization.
repl-backlog-size 25mb

# No of threads
server-threads 16

# Replication
multi-master yes
active-replica yes
replicaof {{ replicaof1 }} {{ keydb_port }} --> EACH NODE HAS THE OTHER 3 AS REPLICASOF
replicaof {{ replicaof2 }} {{ keydb_port }} --> EACH NODE HAS THE OTHER 3 AS REPLICASOF
replicaof {{ replicaof3 }} {{ keydb_port }} --> EACH NODE HAS THE OTHER 3 AS REPLICASOF
```

***Key count on keydb-1 (writer)***
``` shell
$ 127.0.0.1:6379> keys *
...
...
$ 183) "my_last_key" --> 183 KEYS
```
***Key count on keydb-2 (reader)***
``` shell
$ 127.0.0.1:6379> keys *
...
...
$ 180) "my_last_key" --> 180KEYS
```

**Replication information from the nodes**
***keydb-1***
``` shell
$ 127.0.0.1:6379> role
1) 1) "active-replica"
   2) "keydb-2"
   3) (integer) 6379
   4) "connected"
   5) (integer) 2418272672700
2) 1) "active-replica"
   2) "keydb-3"
   3) (integer) 6379
   4) "connected"
   5) (integer) 4287591675913
3) 1) "active-replica"
   2) "keydb-4"
   3) (integer) 6379
   4) "connected"
   5) (integer) 3447852560191
```
***keydb-2***
```shell
127.0.0.1:6379> role
1) 1) "active-replica"
   2) "keydb-1"
   3) (integer) 6379
   4) "connected"
   5) (integer) 2981280733983
2) 1) "active-replica"
   2) "keydb-3"
   3) (integer) 6379
   4) "connected"
   5) (integer) 4287707349480
3) 1) "active-replica"
   2) "keydb-4"
   3) (integer) 6379
   4) "connected"
   5) (integer) 3447938567903
```
***keydb-3***
```shell
127.0.0.1:6379> role
1) 1) "active-replica"
   2) "keydb-2"
   3) (integer) 6379
   4) "connected"
   5) (integer) 2418416826975
2) 1) "active-replica"
   2) "keydb-4"
   3) (integer) 6379
   4) "connected"
   5) (integer) 3447959624778
3) 1) "active-replica"
   2) "keydb-1"
   3) (integer) 6379
   4) "connected"
   5) (integer) 2981309268116
```
***keydb-4***
```shell
127.0.0.1:6379> role
1) 1) "active-replica"
   2) "keydb-2"
   3) (integer) 6379
   4) "connected"
   5) (integer) 2418439958967
2) 1) "active-replica"
   2) "keydb-3"
   3) (integer) 6379
   4) "connected"
   5) (integer) 4287758973088
3) 1) "active-replica"
   2) "keydb-1"
   3) (integer) 6379
   4) "connected"
   5) (integer) 2981332498682
```

***info replication***
```shell
127.0.0.1:6379> info replication
# Replication
role:active-replica
master_global_link_status:up
connected_masters:3
master_host:keydb-2
master_port:6379
master_link_status:up
master_last_io_seconds_ago:0
master_sync_in_progress:0
slave_read_repl_offset:2418456484944
slave_repl_offset:2418456484944
master_1_host:keydb-3
master_1_port:6379
master_1_link_status:up
master_1_last_io_seconds_ago:0
master_1_sync_in_progress:0
slave_read_repl_offset:4287775501994
slave_repl_offset:4287775501994
master_2_host:keydb-4
master_2_port:6379
master_2_link_status:up
master_2_last_io_seconds_ago:0
master_2_sync_in_progress:0
slave_read_repl_offset:3447989710008
slave_repl_offset:3447989710008
slave_priority:100
slave_read_only:0
replica_announced:1
connected_slaves:3
slave0:ip=10.11.0.1,port=6379,state=online,offset=2981347034216,lag=1
slave1:ip=10.11.0.2,port=6379,state=online,offset=2981349023715,lag=0
slave2:ip=10.11.0.3,port=6379,state=online,offset=2981348799785,lag=0
master_failover_state:no-failover
master_replid:9939a8c8472984c8975584d5c540238bac69ce3d
master_replid2:0000000000000000000000000000000000000000
master_repl_offset:2981349050313
second_repl_offset:-1
repl_backlog_active:1
repl_backlog_size:26214400
repl_backlog_first_byte_offset:2981322835914
repl_backlog_histlen:26214400
```
On the 2 remaining standby instances, key count is 183 as well. 

**Expected behavior**
All 4 nodes should have the same data.

**Additional Information**
The discrepancy between the nodes remains for long periods of time, and the keys to the 'inconsistent' node never really catch up, unless the keydb instance is restarted.
If restarted, it replicates to the other masters, but then the issue appears again.

Any advice welcome.
Thank you

**Comment by solracsf** [comment]
Is this behavior persistent on v6.3.1?
https://download.keydb.dev/pkg/

**Comment by ArisPavlides** [comment]
Hi @solracsf 

From our part at least, we have not observed similar behaviour in v6.3.1 but then again we have been running it for less than a day so it may be too early to tell.

Moreover, since 6.3.0 is still pre-release, I am not sure if it's wise to run 6.3.1 on our production clusters.

**Comment by JohnSully** [comment]
6.3.1 Is a bugfix release to resolve major issues in 6.3.0.  As far as stability and what version you should use I want to make a few points to help with that decision:

1. In KeyDB since our major versions track Redis' major versions, the second number is where we indicate major changes.  6.3.0 Integrates what was formerly KeyDB pro code.  Because our open source code has always had more adoption than pro this is likely there will be issues found by the community we haven't observed with past users of Pro.

2. KeyDB Pro includes a number of substantial changes to the core of KeyDB to unlock better performance in the future.  This extra complexity means that 6.3.0 is an even greater change than past "major" updates have been.

Ultimately the pro codebase of which 6.3.0 is based on is the future of KeyDB and is our only branch from now on.  All future work will be based off of 6.3.0's code.  6.2 builds are still available on Docker if you wish to give it more bake time before deploying in your environment.

**Comment by iacovosb** [comment]
Hello,
We have set up keydb v6.3.1 for the last couple of days; have not identified the replication issue as of now.
I will update accordingly after a few more days of running.
Thank you

**Comment by benschermel** [comment]
Hi @iacovosb, has this issue been resolved? If so we will close this issue.

**Comment by iacovosb** [comment]
Hello @benschermel ,
The issue has been resolved in v6.3.1
Thank you

**Comment by benschermel** [comment]
closing issue as resolved
